From f0d703248bb3d53cdaa35c6660c4d55ac08adf39 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:03:36 -0400
Subject: [usb] wusbcore: don't mark WA_SEG_DTI_PENDING segs as done in urb_dequeue

Message-id: <1410887463-20674-145-git-send-email-dzickus@redhat.com>
Patchwork-id: 95291
O-Subject: [RHEL7 PATCH 144/591] usb: wusbcore: don't mark WA_SEG_DTI_PENDING segs as done in urb_dequeue
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 5090ecea133325f762704f00963bca1b024ee691
Author: Thomas Pugliese <thomas.pugliese@gmail.com>
Date:   Tue Mar 4 11:24:56 2014 -0600

    usb: wusbcore: don't mark WA_SEG_DTI_PENDING segs as done in urb_dequeue

    Data for transfer segments in the WA_SEG_DTI_PENDING state is actively
    being read by the driver.  Let the buffer read callback handle the
    transfer cleanup since cleaning it up in wa_urb_dequeue will cause the
    read callback to access invalid memory if the transfer is completed.

    Signed-off-by: Thomas Pugliese <thomas.pugliese@gmail.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/wusbcore/wa-xfer.c b/drivers/usb/wusbcore/wa-xfer.c
index 784f6db..ff8df95 100644
--- a/drivers/usb/wusbcore/wa-xfer.c
+++ b/drivers/usb/wusbcore/wa-xfer.c
@@ -2008,6 +2008,16 @@ int wa_urb_dequeue(struct wahc *wa, struct urb *urb, int status)
   case WA_SEG_ABORTED:
    break;
    /*
+    * The buf_in data for a segment in the
+    * WA_SEG_DTI_PENDING state is actively being read.
+    * Let wa_buf_in_cb handle it since it will be called
+    * and will increment xfer->segs_done.  Cleaning up
+    * here could cause wa_buf_in_cb to access the xfer
+    * after it has been completed/freed.
+    */
+  case WA_SEG_DTI_PENDING:
+   break;
+   /*
     * In the states below, the HWA device already knows
     * about the transfer.  If an abort request was sent,
     * allow the HWA to process it and wait for the
@@ -2016,7 +2026,6 @@ int wa_urb_dequeue(struct wahc *wa, struct urb *urb, int status)
     */
   case WA_SEG_SUBMITTED:
   case WA_SEG_PENDING:
-  case WA_SEG_DTI_PENDING:
    /*
     * Check if the abort was successfully sent.  This could
     * be false if the HWA has been removed but we haven't
-- 
1.7.1